Doc(eos_designs): add precisions regarding eos_designs and eos_cli_config_gen variables (#1743)
aristanetworks/ansible-avd: ansible_collections/arista/avd/roles/eos_designs/doc/custom-structured-configuration.md

## Variables Precedence

_**TL;DR:** eos_cli_config_gen variables in the input files will be overwritten
by the eos_designs generated facts when the role produces a value for the same
key, use custom_strutcured_configuration to make sure the values make it in the intended configurations._

### Overview

AVD `eos_designs` role flow is expected to work as follow:

1. from the input variables (inventory, group_vars, host_vars, ...) the `eos_designs` role generates the "structured_config"
2. The "structured_config" is used as input to the `eos_cli_config_gen` role to generate the intended EOS configuration.

A caveat is that in the inventory, group_vars, host_vars,... you can define some variables that are
used by `eos_designs` and others that are used by `eos_cli_config_gen` roles. And while it is perfectly acceptable,
the behavior can sometimes be confusing. If the key name for an `eos_cli_config_gen` variable in the variable files is the same as one of
the key name of a variable generated by `eos_designs` steps, only the value from `eos_designs` step will be configured.
However if there is no clash, then the value of the variable will be the one defined in the variable files.

> The consequence is that upgrading AVD to a newer version for instance, or configuring additional `eos_designs` functionnality in an existing AVD project,

Reviewer: ```suggestion
The consequence is that upgrading AVD to a newer version for instance, or configuring additional `eos_designs` functionality in an existing AVD project,
```

</div>

The recommendation is to use `custom_structured_configuration` as described below. This configuration will be merged on top of the `eos_designs` generated variables.
There are multiple ways of supplying Custom Structured Configuration and they can all be combined.

### Examples

A good example to demonstrate the various behaviors is to use the `ip name-server` configuration as it is quite straightforward. It is configured by `name_servers` for `eos_designs` and by `name_server` for `eos_cli_config_gen`.

#### eos_designs variables only

```
---
# Only eos_designs name_servers variables
name_servers:
  - 192.168.42.10
  - 192.168.42.40
```

will generate as intended config:

```
ip name-server vrf MGMT 192.168.42.10
ip name-server vrf MGMT 192.168.42.40
```

#### eos_cli_config_gen variables ONLY
```
---
# Only name_server from eos_ci_config_gen.
# The variables will make it to the intended config
name_server:
  source:
    vrf: EOS_CLI
  nodes:
    - 8.8.8.8
    - 4.4.4.4

```

will generate as intended config:

```
ip name-server vrf EOS_CLI 4.4.4.4
ip name-server vrf EOS_CLI 8.8.8.8
```

#### eos_cli_config_gen variables overwritten by eos_designs variables

```

> # Both name_servers from eos_design and name_server from

Reviewer 2: Typo: eos_design

# eos_ci_config_gen. The second ones WON'T be displayed
# as they are overwritten by the generated structured_configuration
name_servers:
  - 192.168.42.10
  - 192.168.42.40

name_server:
  source:
    vrf: EOS_CLI
  nodes:
    - 8.8.8.8
    - 4.4.4.4
```

will generate as intended config:

```
ip name-server vrf MGMT 192.168.42.10
ip name-server vrf MGMT 192.168.42.40
```

#### eos_designs variables overwritten by custom_structured_configuration

```
---

> # Both name_servers from eos_design and leveraging the

Reviewer 2: Same typo.

    peer_interface: Ethernet123
    peer_type: my_precious

> 

Reviewer: remove extra empty line

Author: actually this line (and the one above) should not have been removed.. fixing this

  - 10.10.10.11
  - 10.10.10.12
```

> ## Variables Precedence

Reviewer: I think this description should be moved up above the structured config data model stuff. So the user understands the problem before trying to solve it.

Author: moved on top

_**TL;DR:** eos_cli_config_gen variables in the input files will be overwritten
by the eos_designs generated facts when the role produces a value for the same
key, use custom_strutcured_configuration to make sure the values make it in the intended configurations._

### Overview

AVD `eos_designs` role flow is expected to work as follow:

> 1. from the input variables (inventory, group_vars, host_vars,...), a task first sets ansible facts for each device

Reviewer: For this description I think the flow could be simplified to:

1. from the input variables (inventory, group_vars, host_vars, ...) the `eos_designs` role generates the "structured_config"
2. The "structured_config" is used as input to the `eos_cli_config_gen` role to generate the intended EOS configuration.

Author: addressed in [215e84d](https://github.com/aristanetworks/ansible-avd/pull/1743/commits/215e84d71a018127cdd37073dde4c374dc962906)

2. the facts are then used in a subsequent task to generate a structured_config
3. this structured_config is what is then used by `eos_cli_config_gen` role to generate the intended config

> A caveat is that it is possible to define in the inventory, group_vars, host_vars,... some variables that are

Reviewer: ```suggestion
A caveat is that in the inventory, group_vars, host_vars,... you can define some variables that are
```

used by `eos_designs` and others used by `eos_cli_config_gen` roles. And while it is perfectly acceptable,
the behavior can sometimes be confusing. If the key name for an `eos_cli_config_gen` variable in the variable files is the same as one of
the key name of a variable generated by `eos_designs` steps, only the value from `eos_designs` step will be configured.
However if there is no clash, then the value of the variable will be the one defined in the variable files.

> The consequence is that upgrading AVD to a newer version for instance, or configuring additional `eos_designs` functionnality in some existing AVD repo,

Reviewer: ```suggestion
The consequence is that upgrading AVD to a newer version for instance, or configuring additional `eos_designs` functionnality in an existing AVD project,
```

> could lead to some additional variable keys being generated by `eos_designs` in the structured_config, overwriting previously working `eos_cli_config_gen` keys.

Reviewer: ```suggestion
could lead to additional variable keys being generated by `eos_designs` in the structured_config, overwriting previously working `eos_cli_config_gen` keys.
```

<!-- ![Figure 1: Variables Precedence between roles](../../../media/variables_precedence.png) -->

<div style="text-align:center">
  <img src="../../../media/variables_precedence.png" />
</div>

> The recommendation is to use custom_structured_configuration as described in [the following section of this guide](./custom-structured-configuration.md). This configuration will always prevails over the `eos_designs` generated variables.

Reviewer: After you move this up the vars for custom_structured_configuration will be below on this page.
```suggestion
The recommendation is to use `custom_structured_configuration` as described below. This configuration will be merged on top of the `eos_designs` generated variables.
```
